### PR TITLE
Solver tax implementation

### PR DESCRIPTION
**Pull request by syahirAmali**

### CODE REVIEW

File: contracts/Rewarder.sol [hunk 1]
```diff
@@ -340,64 +374,62 @@ contract Rewarder is IRewarder, Pausable {
             IERC20(token).balanceOf(address(this)) == currentBalance + payment,
             "Invalid token transfer"
         );
+
+        ITaxManager taxManager = getTaxManager();
+        uint256 protocolTaxRate = taxManager.protocolTaxRate();
+        uint256 baseDivisor = taxManager.taxBaseDivisor();
```
Comment (Cosmodude) [inline contracts/Rewarder.sol:380]: check if the protocol tax rate is the right tax rate

Comment (Cosmodude) [reply contracts/Rewarder.sol:380]: it is

File: contracts/Rewarder.sol [hunk 2]
```diff
@@ -340,64 +374,62 @@ contract Rewarder is IRewarder, Pausable {
...
-                rewardReferrers(solverHandler, tax, taxRateDivisor, token);
-                _processPayment(solver, token, solverReward);
+                rewardReferrers(solverHandler, tax, baseDivisor, token);
+                // NOTE: event emits sending reward token to solverhandler, but process payment is sending to the owner of the handler, should probably use the same value here
+                _processPayment(
+                    IReferralHandler(solverHandler).owner(), 
+                    token, 
+                    rewardValue - tax
+                );
+                emit RewardNativeClaimed(solverHandler, msg.sender, rewardValue - tax);
```
Comment (Cosmodude) [inline contracts/Rewarder.sol:403]: cei

File: contracts/Rewarder.sol [hunk 3]
```diff
@@ -340,64 +374,62 @@ contract Rewarder is IRewarder, Pausable {
...
             
-            emit RewardNativeClaimed(solverHandler, msg.sender, solverReward);
-
             {
-                address solver = IReferralHandler(solverHandler).owner();
-
-                rewardReferrers(solverHandler, tax, taxRateDivisor, token);
-                _processPayment(solver, token, solverReward);
+                rewardReferrers(solverHandler, tax, baseDivisor, token);
+                // NOTE: event emits sending reward token to solverhandler, but process payment is sending to the owner of the handler, should probably use the same value here
```
Comment (Cosmodude) [inline contracts/Rewarder.sol:397]: we want to track the rewards by account, not its owner (owner can change)

File: contracts/Rewarder.sol [hunk 4]
```diff
@@ -142,30 +166,44 @@ contract Rewarder is IRewarder, Pausable {
...
         require(
             IERC20(token).balanceOf(address(this)) == currentBalance + amount,
             "Invalid token transfer"
         );
 
         ITaxManager taxManager = getTaxManager();
-        uint256 protocolTaxRate = taxManager.protocolTaxRate();
         uint256 taxRateDivisor = taxManager.taxBaseDivisor();
+        
+        (uint256 referralTax, uint256 platformTax, uint256 platformTreasuryTax) = calculateSolverTax(amount);
```
Comment (Cosmodude) [inline contracts/Rewarder.sol:178]: return values mixed order

File: contracts/TaxManager.sol [hunk 5]
```diff
@@ -15,22 +15,21 @@ contract TaxManager is ITaxManager {
...
 
     // Tax Rates variables
     uint256 public disputeDepositRate;  // with base divisor
-    uint256 public solverTaxRate; // shoulod be 
     uint256 public protocolTaxRate;
 
     // Treasury addresses
-    address public seekerFeesTreasury;
-    address public solverFeesTreasury;
+    address public platformTreasury;
```
Comment (Cosmodude) [inline contracts/TaxManager.sol:32]: both seeker and solver tax will be deposited here, right?

**Review by Cosmodude — CHANGES_REQUESTED** [review]

File: contracts/Rewarder.sol [hunk 6]
```diff
@@ -93,37 +94,61 @@ contract Rewarder is IRewarder, Pausable {
      * @dev Can be called by anyone
      * @param solverId Nft Id of the quest solver
      */
-    // TODO: Currently missing the solver tax, only has protocol tax
     function handleRewardNative(uint32 solverId, uint256 amount) public payable whenNotPaused {
         address escrow = msg.sender;
 
+        // NOTE: check can be griefed and DOSed
```
[the remark this answers is not in the document]
Comment (Cosmodude) [reply contracts/Rewarder.sol:100]: If neither a receive Ether nor a payable fallback function is present, the contract cannot receive Ether through a transaction that does not represent a payable function call and throws an exception.
https://docs.soliditylang.org/en/latest/contracts.html

[the remark this answers is not in the document]
Comment (syahirAmali) [reply contracts/Rewarder.sol:100]: You can basically create a smart contract that has a selfdestruct and force eth into a contract without a receive or fallback function.

https://solidity-by-example.org/hacks/self-destruct/

[the remark this answers is not in the document]
Comment (Cosmodude) [reply contracts/Rewarder.sol:100]: oh, right!
Then will set escrow to send the whole balance, not payment amount

File: contracts/Rewarder.sol [hunk 7]
```diff
@@ -175,27 +213,24 @@ contract Rewarder is IRewarder, Pausable {
         returns (uint256 platformTax_, uint256 referralTax_)
     {
         ITaxManager taxManager = getTaxManager();
-        (platformTax_, referralTax_) = _calculateSeekerTax(taxManager, _paymentAmount);
+        (platformTax_, referralTax_) = TaxCalculator._calculateSeekerTax(taxManager, _paymentAmount);
     }
 
-    function _calculateSeekerTax(ITaxManager _taxManager, uint256 _paymentAmount) 
-        internal 
+    function calculateSolverTax(uint256 _paymentAmount)
```
[the remark this answers is not in the document]
Comment (Cosmodude) [reply contracts/Rewarder.sol:219]: it's a wrapper for lib function

**Review by Cosmodude — APPROVED** [review]
cei and return values order fixed

**Pull request merged**